tracer/src/Datadog.Trace/AspNet/TracingHttpModule.cs

```diff
@@ -182,6 +182,20 @@ private void OnEndRequest(object sender, EventArgs eventArgs)
                             scope.Span.ResourceName = $"{app.Request.HttpMethod.ToUpperInvariant()} {path.ToLowerInvariant()}";
                         }
 
+                        var security = Security.Instance;
+                        if (security.Settings.Enabled)
+                        {
+                            var httpContext = (sender as HttpApplication)?.Context;
+
+                            if (httpContext == null)
+                            {
+                                return;
+                            }
+
+                            security.InstrumentationGateway.RaiseRequestEnd(httpContext, httpContext.Request, scope.Span, null);
+                            security.InstrumentationGateway.RaiseLastChanceToWriteTags(httpContext, scope.Span);
```

Comment:
The ordering here seems wrong, surely `LastChanceToWriteTags` would come before `RequestEnd` ?

Reply:
The idea is to smoothen out differences between core and framework. And usually, after request end, there's other events, like Response completed, response starting to send headers, response starting to send content. Ideally we would like to be at response starting to send headers, but it's not possible. So this event is the very last one we can have with appsec over spans in Framework mode using this module, to be as close as possible to the final response headers and status code. It would be the equivalent of what we had in `HttpTransport.Framework.cs `, `HttContext.AddOnRequestCompleted`, except that it wasn't working, cause this was fired too late compared to the Span lifecycle, way after RequestEnd. 
But for Core, here, we actually do: 
```
 httpContext.Response.OnStarting(() =>
                    {
                        // we subscribe here because in OnHostingHttpRequestInStop or HostingEndRequest it's too late,
                        // the waf is already disposed by the registerfordispose callback
                        security.InstrumentationGateway.**RaiseRequestEnd**(httpContext, request, span);
                        return System.Threading.Tasks.Task.CompletedTask;
                    });

                    httpContext.Response.OnCompleted(() =>
                    {
                        security.InstrumentationGateway.**RaiseLastChanceToWriteTags**(httpContext, span);
                        return System.Threading.Tasks.Task.CompletedTask;
                    });
```
In that case, the LastChance is indeed well after the request end event, on completed, when the Span is not yet disposed (it is disposed in `Hosting.HttpRequestIn.Stop`, thankfully, Core offers a more granular lifecycle).
Ideally we wanna raise the LastChance the latest possible in the lifecycle to read response headers, so after request ends, before response starts sending, but for framework, given when and how Spans are disposed, we raise it right after, and try to mimic the Core way but I agree that it's a bit artificial..

Reply:
Right, sorry, I miss understood, I read `RequestEnd` as `ResponseEnd`.

So, more or less:
- `RequestEnd` means user code has finished executing.
- `RaiseLastChanceToWriteTags` means that response is about to start, so we won't be able to modify the results after this?

Reply:
Well in Core and Framework, it means different things, but for the sake of having one generic event, I just named it RequestEnd, when in framework, it's really RequestEnd, and in Core, it's a callback that Core plays before starting sending headers. This helped me understand the 2 complete different lifecycles 😅 
![image](https://user-images.githubusercontent.com/8353486/148949962-fe56955e-db81-47ba-ba50-468b23968f7e.png)

In core, things are obviously better, because the LastChanceToWriteTags (which for us is used to read response headers so it has to be at the last moment) is really at the very end, once the Response has been sent. Whereas in framework, there's still other later events that makes it theoretically possible to change response headers.. but as the Scope is disposed before, we can't really use them for now.

So, more or less:

- `RequestEnd` means request end in framework and response about to start sending in core. (users can still register callback on those events so we have no garantee that user code has finished executing, but the mvc part should indeed be finished)
-  RaiseLastChanceToWriteTags means  request end in framework but is still after our first request end gateway event and response completed in Core. 

I hope it's not more confusing😅

Reply:
No, that's very clear thank you.